core/thread/src/TReentrantRWLock.cxx

```diff
@@ -221,100 +221,151 @@ void TReentrantRWLock<MutexT, RecurseCountsT>::WriteUnLock(TVirtualRWMutex::Hint
+      fRecurseCounts.fWriteRecurse = typedState.fWriteRecurse - 1;
 
       // Release this thread's write lock
-      WriteUnLock(reinterpret_cast<TVirtualRWMutex::Hint_t *>(pState->fReadersCountLoc));
-   } else if (readerCount) {
-      // Now set the lock for release.
-      fReaders -= (readerCount-1);
-      *(pState->fReadersCountLoc) = 1;
+      WriteUnLock(hint);
+   } else if (pStateDelta->fReadersCountLoc != 0) {
+      // Notify through ReadUnLock()
+      *typedState.fReadersCountLoc = typedState.fReadersCount - 1;
+      fRecurseCounts.fWriteRecurse = typedState.fWriteRecurse;
 
       // Release this thread's reader lock(s)
-      ReadUnLock(reinterpret_cast<TVirtualRWMutex::Hint_t *>(pState->fReadersCountLoc));
+      ReadUnLock(hint);
    }
-
-   // Do something.
-   //::Fatal("Reset()", "Not implemented, contact pcanal@fnal.gov");
-   return std::move(pState);
+   // else earlierState and *this are identical!
+   return pStateDelta;
 }
 
+//////////////////////////////////////////////////////////////////////////
+/// Re-apply a delta.
+
 template <typename MutexT, typename RecurseCountsT>
-void TReentrantRWLock<MutexT, RecurseCountsT>::Restore(std::unique_ptr<TVirtualMutex::State> &&state)
-{
-   TReentrantRWLockState<MutexT, RecurseCountsT> *pState = dynamic_cast<TReentrantRWLockState<MutexT, RecurseCountsT> *>(state.get());
-   if (!pState) {
-      if (state) {
-         SysError("Restore", "LOGIC ERROR - invalid state object!");
-         return;
-      }
-      // No state, do nothing.
+void TReentrantRWLock<MutexT, RecurseCountsT>::Apply(std::unique_ptr<StateDelta> &&state) {
+   if (!state) {
+      Error("TReentrantRWLock::Apply", "Cannot apply empty delta!");
       return;
    }
 
-   // At a restore point, this thread should not be holding any part
-   // of the lock (if it does the following code will forget about it)
-   assert( *(pState->fReadersCountLoc) == 0);
-   // assert( auto local = fRecurseCounts.GetLocal() && fRecurseCounts.IsNotCurrentWriter(local) )
+   using StateDelta_t = TReentrantRWLockStateDelta<MutexT, RecurseCountsT>;
+   const StateDelta_t* typedDelta = static_cast<const StateDelta_t*>(state.get());
 
-   const auto readerCount = pState->fReadersCount;
+   if (typedDelta->fDeltaWriteRecurse < 0) {
+      Error("TReentrantRWLock::Apply", "Negative write recurse count delta!");
+      return;
+   }
+   if (typedDelta->fDeltaReadersCount < 0) {
+      Error("TReentrantRWLock::Apply", "Negative read count delta!");
+      return;
+   }
+   AssertReadCountLocIsFromCurrentThread(typedDelta->fReadersCountLoc);
 
-   if (pState->fIsWriter) {
-      WriteLock();
-      // Now that we go the lock, fix up the recursion count.
-      std::unique_lock<MutexT> lock(fMutex);
-      fRecurseCounts.fWriteRecurse = pState->fWriteRecurse;
-      *(pState->fReadersCountLoc) = readerCount;
-      fReaders +=  readerCount;
-   } else if (readerCount) {
+   if (typedDelta->fDeltaReadersCount != 0) {
```

Comment:
The order seems wrong.   In general we can have a WriteLock with some ReadLocks and in that case we need to acquire the WriteLock.